client/hostvolumemanager/host_volumes.go

```diff
@@ -20,19 +22,59 @@ var (
 type HostVolumeManager struct {
 	pluginDir      string
 	sharedMountDir string
+	stateMgr       HostVolumeStateManager
 
 	log hclog.Logger
 }
 
-func NewHostVolumeManager(logger hclog.Logger, pluginDir, sharedMountDir string) *HostVolumeManager {
+func NewHostVolumeManager(logger hclog.Logger,
+	stateMgr HostVolumeStateManager, restoreTimeout time.Duration,
+	pluginDir, sharedMountDir string) (*HostVolumeManager, error) {
+
 	log := logger.Named("host_volume_mgr")
 
 	// db TODO(1.10.0): how do we define the external mounter plugins? plugin configs?
-	return &HostVolumeManager{
-		log:            log,
+	hvm := &HostVolumeManager{
 		pluginDir:      pluginDir,
 		sharedMountDir: sharedMountDir,
+		stateMgr:       stateMgr,
+		log:            log,
+	}
+
+	if err := hvm.restoreState(stateMgr, restoreTimeout); err != nil {
+		return nil, err
+	}
+
+	return hvm, nil
+}
+
+func (hvm *HostVolumeManager) restoreState(state HostVolumeStateManager, timeout time.Duration) error {
+	vols, err := state.GetDynamicHostVolumes()
+	if err != nil {
+		return err
+	}
+	if len(vols) == 0 {
+		return nil // nothing to do
 	}
+
+	// re-"create" the volumes - plugins have the best knowledge of their
+	// side effects, and they should be idempotent.
+	var wg sync.WaitGroup
+	for _, vol := range vols {
+		wg.Add(1)
+		func() {
+			defer wg.Done()
+			ctx, cancel := context.WithTimeout(context.Background(), timeout)
+			defer cancel()
+			// note: this will rewrite client state that we just restored
+			if _, err := hvm.Create(ctx, vol.CreateReq); err != nil {
+				hvm.log.Error("failed to restore", "volume_id", vol.ID, "error", err)
+				// db TODO: multierror w/ mutex?
```

Comment:
If you're going to collect errors, you might want to send them back out on a channel and collect them that way. That would also let you get rid of the wait group.

Reply:
I went with a [`multierror.Group{}`](https://pkg.go.dev/github.com/hashicorp/go-multierror#Group)